storage/remote/queue_manager.go

```diff
@@ -1232,20 +1240,95 @@ func (s *shards) stop() {
 	}
 }
 
+func (s *shards) updateDroppedMetrics() {
+	// In this case we drop all samples in the buffer and the queue.
+	// Remove them from pending and mark them as failed.
+	droppedSamples := int(s.enqueuedSamples.Load())
+	droppedExemplars := int(s.enqueuedExemplars.Load())
+	droppedHistograms := int(s.enqueuedHistograms.Load())
+	s.qm.metrics.pendingSamples.Sub(float64(droppedSamples))
+	s.qm.metrics.pendingExemplars.Sub(float64(droppedExemplars))
+	s.qm.metrics.pendingHistograms.Sub(float64(droppedHistograms))
+	s.qm.metrics.failedSamplesTotal.Add(float64(droppedSamples))
+	s.qm.metrics.failedExemplarsTotal.Add(float64(droppedExemplars))
+	s.qm.metrics.failedHistogramsTotal.Add(float64(droppedHistograms))
+	s.samplesDroppedOnHardShutdown.Add(uint32(droppedSamples))
+	s.exemplarsDroppedOnHardShutdown.Add(uint32(droppedExemplars))
+	s.histogramsDroppedOnHardShutdown.Add(uint32(droppedHistograms))
+}
+
+func (s *shards) reshard(numShards int) bool {
+	// Exclusive lock to ensure that this does not run concurrently with enqueue.
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	newQueues := make([]*queue, numShards)
+	for i := 0; i < numShards; i++ {
+		newQueues[i] = newQueue(s.qm.cfg.MaxSamplesPerSend, s.qm.cfg.Capacity)
+	}
+
+	successful := s.redistributeSamples(newQueues)
```

Comment:
One thing I'm thinking about, with the way things are written currently if we are downsharding we could possible often get `successful = false` when attempting to redistribute. 

If we're downsharding and all the existing queues are at full `Capacity`, the newer _lower_ amount of shards won't be able to buffer that many samples. In which case, we either want to know that ahead of time and not attempt to redistribute _or_ have a way to ensure queues that are having samples redistributed to them can send when they're full.

Additionally, we're potentially doubling the amount of memory used by enqueued samples for the duration of the resharding operation.